### ISSUE TITLE
ArrayIndexOutOfBounds in when resolving and -runsystempackages contains duplication (related to #1439)

### ISSUE THREAD

**Issue by zhsc** [bndtools/bnd#2264]
When the same package is repeated in -runsystempackages, or if -runsystempackages directive is included twice (i.e. included from another file), an ArrayIndexOutOfBoundsException is thrown.

This appears to be a variant of issue #1439 
I'm using bndtools 3.5.0.REL-20170929-190052-ga0d7a6b.

Here's a sample bndrun file:

```
-runfw: org.apache.felix.framework;version='[5.6.0,5.6.0]'
-runee: JavaSE-1.8
-runrequires: osgi.identity;filter:='(osgi.identity=osgi.configurer.provider)'

-runsystempackages: sun.reflect, sun.reflect
```

Note: the runrequires can be anything, doesn't have to be the bundle in the example.

The main concern with this issue is the error messaging makes it really difficult to determine the cause is the -runsystempackages issue esp. given a large bndrun file.

**Comment by bjhargrave** [comment]
Can you please add the full stacktrace to this issue? Thanks.

**Comment by zhsc** [comment]
Hi,
Nothing is logged to the Eclipse error log. The only message is in the resolution result dialog.

I've also tried adding -consoleLog and -debug to eclipse startup, but nothing else is logged.

Is there a way I can get to the stack trace of the error?  

![image](https://user-images.githubusercontent.com/6600526/34882868-e1440d02-f785-11e7-8b01-0c0cf53a1c36.png)

**Comment by bjhargrave** [comment]
I am thinking this should really be a fix in the resolver. It should be able to handle duplicated capabilities. @tjwatson, should the resolver be able to handle this?

**Comment by tjwatson** [comment]
Is it possible this is related to https://issues.apache.org/jira/browse/FELIX-5263

Are the two capabilities representing the sun.reflect package identity equals?  The resolver expects the two capabilities to NOT equal each other since they are supposed to be distinct capabilities.

**Comment by bjhargrave** [comment]
Yes, they would be equal. There is nothing I see in `Resource` the would prohibit identical capabilities in the list of capabilities. 

Where do you think it states they must be distinct?

**Comment by tjwatson** [comment]
The spec does say that the same package name can be exported multiple times.  It says this:

Multiple export definitions for the same package are allowed for example, when different attributes are needed for different importers.

I'll agree, not normative text, but gives the motivation for why a bundle would export packages multiple times.  Because they need different attributes.  Also, there is the javadoc of:

https://osgi.org/javadoc/r6/cmpn/org/osgi/service/resolver/ResolveContext.html#findProviders(org.osgi.resource.Requirement)

This returns a list of capabilities in priority order.  I'm not sure how to interpret priority order if the same capability is allowed to be included in the list multiple times.  Say you have capabilities 1,2,3,4 and the list has [4,1,2,3,4].  Is capability 4 considered to be the highest priority or the lowest?

**Comment by bjhargrave** [comment]
Priority order just means there is an order (since it is a List) and that earlier items should be chose over later items. Yes, having 4 in the list twice is awkward but does not appear to be disqualifying under the specs. 4 has a higher priority than 1, 2, or 3 since it comes earlier in list.

It would seem the Resolver is demanding more than the specs require.

**Comment by rickhall** [comment]
If I recall, I think the rule was no duplicates but we realized we needed to relax that slightly for exports so they could have different attributes. However, I don't believe that relaxation was to allow actual duplicate exports, but rather exports of the same package with different attributes (which is technically not a duplicate).

So, the spec may not explicitly state it, but I think that was the intent.
